wayland-client/src/conn.rs

```diff
@@ -44,48 +46,61 @@ impl Connection {
     ///
     /// This is the standard way to initialize a Wayland connection.
     pub fn connect_to_env() -> Result<Self, ConnectError> {
-        let stream = if let Ok(txt) = env::var("WAYLAND_SOCKET") {
-            // We should connect to the provided WAYLAND_SOCKET
-            let fd = txt.parse::<i32>().map_err(|_| ConnectError::InvalidFd)?;
+        // Don't try to connect to `WAYLAND_SOCKET` when it's value is empty.
+        if let Some(socket) = env::var("WAYLAND_SOCKET").ok().filter(|var| !var.is_empty()) {
+            let fd = socket.parse::<i32>().map_err(|_| ConnectError::InvalidFd)?;
             let fd = unsafe { OwnedFd::from_raw_fd(fd) };
-            // remove the variable so any child processes don't see it
+
+            // Remove the variable so any child processes don't see it.
             env::remove_var("WAYLAND_SOCKET");
-            // set the CLOEXEC flag on this FD
-            let flags = rustix::io::fcntl_getfd(&fd);
-            let result = flags
-                .map(|f| f | rustix::io::FdFlags::CLOEXEC)
-                .and_then(|f| rustix::io::fcntl_setfd(&fd, f));
-            match result {
-                Ok(_) => {
-                    // setting the O_CLOEXEC worked
-                    UnixStream::from(fd)
-                }
-                Err(_) => {
-                    // something went wrong in F_GETFD or F_SETFD
-                    return Err(ConnectError::InvalidFd);
-                }
+
+            Self::connect_to_wayland_socket(fd)
+        } else {
+            Self::connect_to_wayland_display()
+        }
+    }
+
+    /// Try to connect to the socket.
+    fn connect_to_wayland_socket(fd: OwnedFd) -> Result<Self, ConnectError> {
+        // Set the CLOEXEC flag on this FD.
+        let flags = rustix_io::fcntl_getfd(&fd);
+        let stream = match flags.and_then(|f| rustix_io::fcntl_setfd(&fd, f | FdFlags::CLOEXEC)) {
+            Ok(_) => {
+                // Setting the O_CLOEXEC worked.
+                UnixStream::from(fd)
             }
+            Err(_) => {
+                // Something went wrong in F_GETFD or F_SETFD.
+                return Err(ConnectError::InvalidFd);
+            }
+        };
+
+        let backend = Backend::connect(stream).map_err(|_| ConnectError::NoWaylandLib)?;
+        Ok(Self { backend })
+    }
+
+    /// Try to connect with WAYLAND_DISPLAY env.
+    pub fn connect_to_wayland_display() -> Result<Self, ConnectError> {
```

Comment:
I question whether we want this to be public, as if the server is giving you an explicit socket, you should probably use it. Or is this a case where winit may want to fallback if the socket is garbage?

Otherwise the name is a possibly confusing, someone may think you use this function to connect to a display, but not realize their application should just use the env.

Reply:
Oh, it should be private.